crates/engine/tree/src/tree/cached_state.rs

```diff
@@ -0,0 +1,296 @@
+//! Implements a state provider that has a shared cache in front of it.
+#![allow(dead_code)]
```

Comment:
Is this not going to be hooked up in this PR?

```suggestion

```

Reply:
eh might as well integrate into execution